File: java/src/main/java/P052.java
import java.util.Arrays;
import java.util.stream.IntStream;

@SuppressWarnings("unused")
public class P052 implements Puzzle {
    @Override
    public String solve() {
        return "" + IntStream.range(1, 1000000)
                .filter(i -> same(i, i * 2))
                .filter(i -> same(i * 3, i * 4))
                .filter(i -> same(i * 5, i * 6))
                .filter(i -> same(i, i * 6))
                .findFirst()
                .orElseThrow(() -> new IllegalArgumentException("no result"));

    }

    private static boolean same(int a, int b) {
        char[] aChars = String.valueOf(a).toCharArray();
        char[] bChars = String.valueOf(b).toCharArray();

        Arrays.sort(aChars);
        Arrays.sort(bChars);
        return Arrays.equals(aChars, bChars);
    }
}

File: erlang/src/p052.erl
-module(p052).

-export([start/0]).

f() ->
  g(1).

g(N) ->
  R = same(N, N * 2) andalso same(N * 3, N * 4) andalso same(N * 5, N * 6) andalso same(N, N * 6),
  if
    R -> N;
    true -> g(N + 1)
  end.

same(A, B) ->
  lists:sort(integer_to_list(A)) == lists:sort(integer_to_list(B)).

start() ->
  io:format("~p ~n", [f()]).